React/forms.md

```diff
@@ -2,24 +2,280 @@
+
+![passing a callback through props](./images/passing-callback.png)
+
+First adding a callback function to `StudentCollection` and passing that function to `NewStudentForm` as a prop.
+
+```javascript
+// student_collection.js
+...
+// callback function to add students to the list
+addStudent = (student) => {
+    const students = this.state.students;
+
+    students.push(student);
+    this.setState({
+      students
+    });
+  }
+
+render() {
+  let studentComponents = this.state.students.map(function(student) {
+    return <Student key={ student.name } name={ student.name } email={ student.email }/>
+  });
+
+  return (
+    <div>
+      <h3>Students</h3>
+      { studentComponents }
+      <NewStudentForm addStudent={this.addStudent} />
+    </div>
+  )
+}
+...
+```
+
+Then we can update the `handleFormSubmit` function.
+
+```javascript
+// new_student_form.js
+...
+handleFormSubmit = (event) => {
+  event.preventDefault();
+  const newStudent = {
+    name: this.state.name,
+    email: this.state.email,
+  };
+
+  this.setState({
+    name: '',
+    email: '',
+  });
+
+  this.props.addStudent(student);
+}
+```
+
+Now if we test the app we should now be able to add students to the list.  However there are some issues.
+1.  We can add students with no names
+2.  We can add students with invalid email addresses
+
+**Exercise**  Update the app to prevent form submission if the name is blank, or the email field is invalid.
+
+## Final Note: Prop Validation
+
+Your linter may complain about the `addStudent` prop in `new_student_form.js` due to, "Prevent missing props validation in a React component definition (react/prop-types)".  It is asking you to add some validation of the props `NewStudentForm` requires.
+
+You can add this code to the `NewStudent` class.
+
+```javascript
+// new_student_form.js
+import PropTypes from 'prop-types';
+
+...
+class NewStudentForm extends Component {
+
+  static propTypes = {
+    addStudent: PropTypes.func.isRequired,
+  }
+...
+```
+
+This will prevent the `NewStudentForm` component from being used without the `addStudent` function as a prop.
 
 ## Vocabulary
 | Term     | Definition     |
 | :------------- | :------------- |
-| Item One       | Item Two       |
+| **Controlled Components**       | Components whos state is managed by a React component.       |
+| **Uncontrolled Components**   |  Components who have their state stored in the DOM. |
+
 
 ## Key Takeaway
-Description of one or two sentences with key takeaway.
+Forms in a React application can either be controlled or uncontrolled.  Controlled components provide a number of capabilities beyound regular HTML forms.  Uncontrolled components can be easier to migrate from a non-react application.  Forms can be used to collect data and pass it back to containers via callback functions in the component's props.
```

Comment:
I'm not sure that the difference between the two is what they'll take away from this lecture.